File: sync-service/ui_state.py
import json
import os
from pathlib import Path

ALLOWED_UI_STATE_KEYS = frozenset(
    {
        "calendarCollapsed",
        "calendarPosition",
        "calendarPositionLocked",
        "calendarSize",
        "notesCollapsed",
        "notesHideCompleted",
        "notesPosition",
        "notesPositionLocked",
        "notesSelectedFile",
        "notesSize",
        "wallpaperPrefs",
        "gadgetVisibility",
        "launcherPosition",
        "clockPosition",
        "clockSize",
        "clockPrefs",
        "pomodoroPosition",
        "pomodoroPrefs",
        "linksPosition",
        "linksSize",
        "linksData",
        "mediaPosition",
        "mediaSize",
        "monitorPosition",
        "monitorSize",
    }
)


def ui_state_path(cache_dir: Path) -> Path:
    return cache_dir / "ui-state.json"


def read_ui_state(cache_dir: Path) -> dict[str, str]:
    path = ui_state_path(cache_dir)
    if not path.exists():
        return {}

    try:
        with open(path, encoding="utf-8") as handle:
            raw = json.load(handle)
    except (OSError, json.JSONDecodeError):
        return {}

    if not isinstance(raw, dict):
        return {}

    values: dict[str, str] = {}
    for key, value in raw.items():
        if key not in ALLOWED_UI_STATE_KEYS:
            continue
        if value is None:
            continue
        values[key] = value if isinstance(value, str) else str(value)
    return values


def merge_ui_state(cache_dir: Path, updates: dict[str, str]) -> dict[str, str]:
    current = read_ui_state(cache_dir)
    for key, value in updates.items():
        if key not in ALLOWED_UI_STATE_KEYS:
            continue
        if value is None:
            current.pop(key, None)
            continue
        current[key] = value if isinstance(value, str) else str(value)

    cache_dir.mkdir(parents=True, exist_ok=True)
    path = ui_state_path(cache_dir)
    tmp_path = path.with_suffix(".tmp")
    with open(tmp_path, "w", encoding="utf-8") as handle:
        json.dump(current, handle, ensure_ascii=False, indent=2)
    os.replace(tmp_path, path)
    return current

File: sync-service-rs/src/ui_state.rs
use std::collections::HashMap;
use std::fs;
use std::path::Path;

const ALLOWED_UI_STATE_KEYS: &[&str] = &[
    "calendarCollapsed",
    "calendarPosition",
    "calendarPositionLocked",
    "calendarSize",
    "notesCollapsed",
    "notesHideCompleted",
    "notesPosition",
    "notesPositionLocked",
    "notesSelectedFile",
    "notesSize",
    "wallpaperPrefs",
    "gadgetVisibility",
    "launcherPosition",
    "clockPosition",
    "clockSize",
    "clockPrefs",
    "pomodoroPosition",
    "pomodoroPrefs",
    "linksPosition",
    "linksSize",
    "linksData",
    "mediaPosition",
    "mediaSize",
    "monitorPosition",
    "monitorSize",
];

fn ui_state_path(cache_dir: &Path) -> std::path::PathBuf {
    cache_dir.join("ui-state.json")
}

fn is_allowed(key: &str) -> bool {
    ALLOWED_UI_STATE_KEYS.contains(&key)
}

pub fn read_ui_state(cache_dir: &Path) -> HashMap<String, String> {
    let path = ui_state_path(cache_dir);
    let Ok(text) = fs::read_to_string(path) else {
        return HashMap::new();
    };
    let Ok(raw) = serde_json::from_str::<serde_json::Value>(&text) else {
        return HashMap::new();
    };
    let Some(obj) = raw.as_object() else {
        return HashMap::new();
    };

    let mut values = HashMap::new();
    for (key, value) in obj {
        if !is_allowed(key) || value.is_null() {
            continue;
        }
        let as_str = match value {
            serde_json::Value::String(s) => s.clone(),
            other => other.to_string(),
        };
        values.insert(key.clone(), as_str);
    }
    values
}

pub fn merge_ui_state(
    cache_dir: &Path,
    updates: &HashMap<String, String>,
) -> anyhow::Result<HashMap<String, String>> {
    let mut current = read_ui_state(cache_dir);
    for (key, value) in updates {
        if !is_allowed(key) {
            continue;
        }
        current.insert(key.clone(), value.clone());
    }

    fs::create_dir_all(cache_dir)?;
    let path = ui_state_path(cache_dir);
    let tmp = path.with_extension("tmp");
    fs::write(&tmp, serde_json::to_string_pretty(&current)?)?;
    fs::rename(tmp, path)?;
    Ok(current)
}
